cpp/src/reductions/minmax.cu

```diff
@@ -0,0 +1,183 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// The translation unit for reduction `minmax`
+
+#include <thrust/transform_reduce.h>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/utilities/device_operators.cuh>
+#include <cudf/reduction.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <type_traits>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+/**
+ * @brief stores the minimum and maximum
+ * values that have been encountered so far
+ *
+ */
+template <typename T>
+struct minmax_pair {
+  T min_val;
+  T max_val;
+  bool min_valid;
+  bool max_valid;
+
+  __host__ __device__ minmax_pair()
+    : min_val(cudf::DeviceMin::identity<T>()),
+      max_val(cudf::DeviceMax::identity<T>()),
+      min_valid(false),
+      max_valid(false){};
+  __host__ __device__ minmax_pair(T val, bool valid_)
+    : min_val(val), max_val(val), min_valid(valid_), max_valid(valid_){};
+  __host__ __device__ minmax_pair(T min_val_, bool min_valid_, T max_val_, bool max_valid_)
+    : min_val(min_val_), max_val(max_val_), min_valid(min_valid_), max_valid(max_valid_){};
+};
+
+/**
+ * @brief functor that accepts two minmax_pairs and returns a
+ * minmax_pair whose minimum and maximum values are the min() and max()
+ * respectively of the minimums and maximums of the input pairs. Respects
+ * validity.
+ *
+ */
+template <typename T>
+struct minmax_with_null_binary_op
+  : public thrust::binary_function<minmax_pair<T>, minmax_pair<T>, minmax_pair<T>> {
+  __host__ __device__ minmax_pair<T> operator()(const minmax_pair<T> &x,
+                                                const minmax_pair<T> &y) const
+  {
+    T x_min = x.min_valid ? x.min_val : cudf::DeviceMin::identity<T>();
+    T y_min = y.min_valid ? y.min_val : cudf::DeviceMin::identity<T>();
+    T x_max = x.max_valid ? x.max_val : cudf::DeviceMax::identity<T>();
+    T y_max = y.max_valid ? y.max_val : cudf::DeviceMax::identity<T>();
+
+    // The only invalid situation is if we compare two invalid values.
+    // Otherwise, we are certain to select a valid value due to the
+    // identity functions above changing the comparison value.
+    bool valid_min_result = x.min_valid || y.min_valid;
+    bool valid_max_result = x.max_valid || y.max_valid;
+
+    return minmax_pair<T>{
+      thrust::min(x_min, y_min), valid_min_result, thrust::max(x_max, y_max), valid_max_result};
```

Comment:
`thrust::min` and `thrust::max` are not specialized for cuda, afaict. Use `min` and `max` intrinsics instead.

Reply:
I just realized this is `__host__ __device__`, and so `min`/`max` are not available in the `__host__` case. Still, the intrinsics are faster and would be preferable over the branching, I think.

Reply:
I wouldn't bother in this usecase. This is certainly going to be bandwidth bound where the computational savings aren't going to be noticeable.